### ISSUE TITLE
gedit gsettings schema not reachable

### ISSUE THREAD

**Issue by tiramiseb** [NixOS/nixpkgs#33277]
## Issue description

```
$ gsettings list-schemas | grep -i gedit
$ 
```

I have also notices that `/nix/store/k5zrf5gz97938mymfsawgfncyyv1caly-gedit-3.22.1/share` is not in the `XDG_DATA_DIRS` environment variable, it may be the cause of all this.

I started a discussion about a similar topic on the mailing list, however please note this is not the same. I am now running NixOS unstable. On 17.09, the gsettings schemas were not overriden correctly. Here, the schema for gedit is "not there". The other schema I had problem with (Tilix) is now okay. This issue looks specific to gedit.

### Steps to reproduce

Install gedit. Try the aforementioned command.

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.73, NixOS, 18.03.git.0bbf671 (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.16`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by jtojnar** [comment]
As I mentioned in the mailing list:

> The problem is, gsettings (or dconf-editor) expect the schemas in some concrete locations. That obviously fails with Nix, where the schemas are spread throughout the nix store. You can provide the location manually
>
>     env XDG_DATA_DIRS=/nix/store/k5zrf5gz97938mymfsawgfncyyv1caly-gedit-3.22.1/share/gsettings-schemas/gedit-3.22.1
>     gsettings list-schemas
>
> but it is not very convenient. Alternately, you could use a script like this:
>
> ```bash
> #!/usr/bin/env nix-shell
> #!nix-shell -i bash -p glib.dev
> schemas=""
> for p in $NIX_PROFILES; do
>     if [[ -d "$p" ]]; then
>         for d in $(nix-store --query --references "$p"); do
>             schemas_dir=$(echo "$d"/share/gsettings-schemas/*/glib-2.0/schemas)
>             if [[ -d "$schemas_dir" ]]; then
>                 schemas="$schemas${schemas:+:}$schemas_dir"
>             fi
>         done
>     fi
> done
> env XDG_DATA_DIRS=$schemas gsettings "$@"
> ```

As fair as I know, NixOS does not add the schema paths to environment’s `XDG_DATA_DIRS`. If you see any in the `gsettings list-schemas` command, they probably come from `services.xserver.desktopManager.gnome3.extraGSettingsOverridePackages` option.

**Comment by tiramiseb** [comment]
```
$ gsettings list-schemas | wc -l
99
```

I can assure you I didn't add 99 schemas in `services.xserver.desktopManager.gnome3.extraGSettingsOverridePackage`. There were already plenty of schemas right after installing the system.

The `XDG_DATA_DIRS` contains 46 entries, you can find its content at https://pastebin.com/PvS7VwKH. There are many directories mentioned in this var and many schemas seen by gsettings, but not all of them are listed. Notably, gedit and nautilus are absent from the list. The Tilix schema, for instance, has been in the list since it was installed.

It is why I think you are wrong wen you write "NixOS does not add the schema paths to environment’s `XDG_DATA_DIRS`". But I may be wrong thinking you are wrong. I'm puzzled because your explanations do not fit what I am experiencing...



When I add the gedit package to extraGSettingsOverridePackage, gsettings is able to see its schemas, meaning that I can customize its settings globally (which is okay for me), also meaning that the bug mentioned in the mailing list seems to affect 17.09. But when we do not explicitly override the settings globally, the schemas are not available and the `gsettings set` is not usable as a user for the missing schemas (I don't care because I wanna configure my computer globally, but others certainly would like to allow their users to customize settings with the gsettings command).

**Comment by jtojnar** [comment]
You are right, I can reproduce https://github.com/NixOS/nixpkgs/issues/33219 with bash.

**Comment by andersk** [comment]
Still an issue. `mutter` is another missing schema; there may be more, I just happened to run into this one.

I found this workaround:

```nix
  services.xserver.desktopManager.gnome3.sessionPath =
    [ pkgs.gnome3.gedit pkgs.gnome3.mutter ];
```

**Comment by piegamesde** [comment]
Still relevant to me.

**Comment by andersk** [comment]
Still an issue.

**Comment by guttermonk** [comment]
I'm having the same issue with Nemo file manager. The missing schema causes it to be unable to disable desktop icons, which is needed to remove the "Desktop" entry from the sidebar. See here for more info: [https://discourse.nixos.org/t/how-to-remove-desktop-from-the-nemo-sidebar/46571](https://discourse.nixos.org/t/how-to-remove-desktop-from-the-nemo-sidebar/46571)

Is there a workaround to this? Please explain like I'm 5. Thank you.

**Comment by Nydragon** [comment]
`gsettings list-schemas | wc -l` returns 0, I am on Hyprland but using some Gnome applications (lollypop, tagger, to name some). Is there a convenient way to automatically enable all of them ?

**Comment by Nydragon** [comment]
> I'm having the same issue with Nemo file manager. The missing schema causes it to be unable to disable desktop icons, which is needed to remove the "Desktop" entry from the sidebar. See here for more info: https://discourse.nixos.org/t/how-to-remove-desktop-from-the-nemo-sidebar/46571
> 
> Is there a workaround to this? Please explain like I'm 5. Thank you.

Programs look at the XDG_DATA_DIRS env var to figure out where to look for schemas, and thus which ones are installed, so you could try this:

```
environment.sessionVariables = {
   XDG_DATA_DIRS = [ "{pkgs.nemo}/share/gsettings-schemas/nemo-something" ];
}
```

Nemo package and path might be wrong, you'll have to check

**Comment by athyfr** [comment]
Putty also seems to have missing GSettings schemas.

It complains about it and crashes when you press 'browse' (for private key file)

**Comment by athyfr** [comment]
YT-DLG has missing GSettings schema.
